File: src/CPP/src/leetcode/RemoveDuplicatesFromSortedListII.cpp
/*
 * [Source] https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
 * [Difficulty]: Medium
 * [Tag]: Linked List
 */

#include <iostream>

using namespace std;

// [Solution]: Keep a prev pointer. If Next several nodes are same, skip.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/linkedlist/RemoveAllDuplicatesFromSortedList.java
 */

/* Java solution
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
		if(head==null || head.next==null) return head;
		HashSet<Integer> set=new HashSet<Integer>();
		ListNode dummy=new ListNode(0);
		ListNode cur=dummy;
		ListNode fast=null;
		while(head!=null){
			if(!set.contains(head.val)){
				int key=head.val;
				set.add(key);
				fast=head.next;
				boolean get=false;
				while(fast!=null){
					if(fast.val==head.val){
						get=true;
						break;
					}
					fast=fast.next;
				}
				if(get){
					cur.next=head.next;
				}else{
					cur.next=head;
					cur=head;
				}
			}else{
				cur.next=head.next;
			}
			head=head.next;
		}
		return dummy.next;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}

File: wwwroot/js/remove-duplicates-from-sorted-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if (head === null) {
        return head;
    }

    let dummy = new ListNode();
    let pre = dummy;
    let cur = head;
    while (cur !== null) {
        let val = cur.val;
        let run = cur.next;
        while (run !== null && run.val === val) {
            run = run.next;
        }

        if (run === cur.next) {
            pre.next = cur;
            pre = cur;
            cur = cur.next;
        } else {
            pre.next = null;
            cur = run;
        }
    }

    return dummy.next;
};

document.getElementById("remove-duplicates-from-sorted-list-ii").innerHTML =
    traverseListNode(deleteDuplicates(createList([0,1,1,1,2,2])));